pkg/volume/util/operationexecutor/operation_generator_test.go

```diff
@@ -0,0 +1,202 @@
+
+		pdName := "pd-volume"
+		podName := "pod-1"
+		pod := getTestPodWithGCEPD(podName, pdName)
+		volumeToUnmount := getTestVolumeToUnmount(podName, pdName, pod)
+		expectedPluginName := tc.expectedPluginName
+
+		if tc.isCsiMigrationEnabled {
+			csiSpec, err := translateSpec(volumeToUnmount.VolumeSpec)
+			if err != nil {
+				t.Fatalf("Can't translate volume to CSI")
+			}
+			_, mapper_error := plugin.(volume.BlockVolumePlugin).NewBlockVolumeMapper(csiSpec, pod, volume.VolumeOptions{})
+			if mapper_error != nil {
+				t.Fatalf("mapper error: %v\n", mapper_error)
+			}
+		}
+
+		unmapVolumeFunc, e := operationGenerator.GenerateUnmapVolumeFunc(volumeToUnmount, nil)
+		if e != nil {
+			t.Fatalf("Error occured while generating unmapVolumeFunc: %v", e)
+		}
+
+		var ee error
+		unmapVolumeFunc.CompleteFunc(&ee)
+
+		metric_family_name := "storage_operation_status_count"
+		st_mf := getMetricFamily(metric_family_name)
+		if st_mf == nil {
+			t.Fatalf("Couldn't find the metric family with name: %s", metric_family_name)
+		}
+
+		labels := make(map[string]string)
+		for _, labelPair := range st_mf.GetMetric()[0].Label {
+			labels[*labelPair.Name] = *labelPair.Value
+		}
+		assert.Equal(t, "success", labels["status"])
+		assert.Equal(t, "unmap_volume", labels["operation_name"])
+		assert.Equal(t, expectedPluginName, labels["volume_plugin"])
+		assert.Equal(t, float64(1), *st_mf.GetMetric()[0].Counter.Value)
+	}
+}
+
+func getTestOperationGenerator(volumePluginMgr *volume.VolumePluginMgr) OperationGenerator {
+	fakeKubeClient := controllervolumetesting.CreateTestClient()
+	fakeRecorder := &record.FakeRecorder{}
+	fakeHandler := volumetesting.NewBlockVolumePathHandler()
+	operationGenerator := NewOperationGenerator(
+		fakeKubeClient,
+		volumePluginMgr,
+		fakeRecorder,
+		false, /* checkNodeCapabilitiesBeforeMount */
+		fakeHandler)
+	return operationGenerator
+}
+
+func getTestVolumeToUnmount(podName string, pdName string, pod *v1.Pod) MountedVolume {
+	volumeMode := v1.PersistentVolumeBlock
+	volumeSpec := &volume.Spec{
+		PersistentVolume: &v1.PersistentVolume{
+			Spec: v1.PersistentVolumeSpec{
+				VolumeMode: &volumeMode,
+				PersistentVolumeSource: v1.PersistentVolumeSource{
+					GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+						PDName: "pdvol-1",
+					},
+				}}}}
+	volumeToUnmount := MountedVolume{
+		PodName:    volumetypes.UniquePodName(podName),
+		VolumeName: v1.UniqueVolumeName(pdName),
+		PodUID:     pod.UID,
+		PluginName: plugins.GCEPDInTreePluginName,
+		VolumeSpec: volumeSpec,
+	}
+	return volumeToUnmount
+}
+
+func getMetricFamily(metric_family_name string) *io_prometheus_client.MetricFamily {
+	metric_families, _ := prometheus.DefaultGatherer.Gather()
+	for _, mf := range metric_families {
+		if *mf.Name == metric_family_name {
+			return mf
+		}
+	}
+	return nil
+}
+
+func initTestPlugins(t *testing.T) (*volume.VolumePluginMgr, volume.VolumePlugin, string) {
```

Comment:
theres a function in `csi_plugin_test.go` `newTestPlugin` that looks very similar. Is there anyway we could reuse that or refactor that into something that could be usable by both tests?

Reply:
I moved the method `csi.newTestPlugin` to a `testing_helper` to reuse it.